Reorient and resample manual labelling

PaulBautin opened the pull request:
Currently label_if_does_not_exist() in process_data.sh does not pick-up the https://github.com/spine-generic/data-multi-subject manual labelling in derivatives.
I tried to implement @jcohenadad suggestion https://github.com/sct-pipeline/csa-atrophy/issues/59#issuecomment-683036793 but this failed, see error:


```
sct_image -i /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-amu01/anat/sub-amu01_T1w_labels-disc-manual.nii.gz -setorient RPI -o /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-amu01/anat/sub-amu01_T1w_RPI_labels-disc-manual.nii.gz

--
Spinal Cord Toolbox (git-master-a52f725fc440edae1d1cbd687ca8c0443a25a6bc)

/home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-amu01/anat/sub-amu01_T1w_labels-disc-manual.nii.gz
Error: make sure /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-amu01/anat/sub-amu01_T1w_labels-disc-manual.nii.gz is an image.
Traceback (most recent call last):
  File "/home/paul/Github/spinalcordtoolbox/scripts/sct_image.py", line 663, in <module>
    main()
  File "/home/paul/Github/spinalcordtoolbox/scripts/sct_image.py", line 281, in main
    im_in = Image(fname_in[0])
  File "/home/paul/Github/spinalcordtoolbox/spinalcordtoolbox/image.py", line 247, in __init__
    self.loadFromPath(param, verbose)
  File "/home/paul/Github/spinalcordtoolbox/spinalcordtoolbox/image.py", line 353, in loadFromPath
    self.data = self.im_file.get_data()
AttributeError: 'NoneType' object has no attribute 'get_data'

```

I was not able to open image manual correction in fsleyes, see overlay error:
```
An error occurred loading the image /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-amu01/anat/sub-amu01_T1w_labels-disc-manual.nii.gz

Details: zran_read returned error (-3)
```
@jcohenadad, @lrouhier  any idea? I have no problem reorienting and resampling the manual labelling i had done with csa-atrophy manual_correction.py.

FIX #59

jcohenadad commented:
@PaulBautin could you please upload the example data you used in the syntax above so we can reproduce and help you find a fix for it?

PaulBautin commented:
This is the new manual labelling on which i cannot resample and reorient: 
[sub-amu01_T1w_labels-disc-manual.zip](https://github.com/sct-pipeline/csa-atrophy/files/5144179/sub-amu01_T1w_labels-disc-manual.zip)

This is the old manual correction i had done with csa-atrophy manual_correction.py (no problem for resampling and reorienting):
[sub-amu01_T1w_RPI_r_labels-disc-manual.zip](https://github.com/sct-pipeline/csa-atrophy/files/5144184/sub-amu01_T1w_RPI_r_labels-disc-manual.zip)

jcohenadad commented:
~~~
julien-macbook:~/Desktop $ fslhd sub-amu01_T1w_labels-disc-manual.nii.gz
Error: file does not appear to be a valid NIFTI image
~~~

your NIFTI file is corrupted-- what did you do with it?

jcohenadad commented:
ah! it's only 104 bits 😉 
~~~
(py37) julien-macbook:~/Desktop $ ll sub-amu01_T1w_labels-disc-manual.nii.gz
-rw-rw-r--@ 1 julien  staff  104 28 Aug 12:39 sub-amu01_T1w_labels-disc-manual.nii.gz
~~~

you probably used the git-annex SHA256 instead of the physical data

jcohenadad commented:
@PaulBautin did you make sure to not fetch data from the latest state of the git-annex repos (because of https://github.com/spine-generic/data-multi-subject/issues/24)

PaulBautin commented:
@jcohenadad, I `git pull` the spine-generic/data-multi-subject dataset this morning after commit https://github.com/spine-generic/data-multi-subject/commit/6ffbb5ca5f83613d5fde546eac2d4c2d9d9cccd8. I did not not `git annex get`.

jcohenadad commented:
`git pull` does not work for *.nii.gz in this git-annex repository (otherwise you are only getting the SHA files). `git-annex` should be used, but as mentioned in https://github.com/sct-pipeline/csa-atrophy/pull/60#issuecomment-683153669 there is currently a problem, so i suggest you just manually grab the file from Lucas's PR.

PaulBautin commented:
Ok, i have tested last commit https://github.com/sct-pipeline/csa-atrophy/pull/60/commits/f17bd775917e6be4e3ee2e22ea839ecd1455973f. @jcohenadad, not sure i addressed your concerns from https://github.com/sct-pipeline/csa-atrophy/issues/59#issuecomment-683036793.

jcohenadad commented on process_data.sh:
```diff
@@ -115,11 +115,24 @@ elif [ $contrast == "t1" ]; then
+# Reorient image to RPI
 sct_image -i ${file}.nii.gz -setorient RPI -o ${file}_RPI.nii.gz
+# Reorient labelling to RPI if manual labelling without reorienting exists
+if [ -e ${file_manual_labels}.nii.gz ]; then
```
if you act on this file now, you probably need to update `label_if_does_not_exist()`

PaulBautin replied:
The process did not work because `label_if_does_not_exist()` was looking for a nonexistent reoriented and resampled manual labelling correction. Here is the processing log file for a subject with a manual segmentation and the new manual labeling:

Found manual labelling without reorientation or resampling:
```
[' -e /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-mniS05/anat/sub-mniS05_T2w_labels-disc-manual.nii.gz ']'
+ echo 'Manual label exists: reorienting'
Manual label exists: reorienting
```
Found reoriented manual labelling:
```
'[' -e /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-mniS05/anat/sub-mniS05_T2w_RPI_labels-disc-manual.nii.gz ']'
+ echo 'Manual label exists: resampling'
Manual label exists: resampling
```
Found manual segmentation:
```
[' -e /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-mniS05/anat/sub-mniS05_T2w_RPI_r_seg-manual.nii.gz ']'
+ echo 'Found! Using manual segmentation.'
Found! Using manual segmentation.
```
Found manual labelling:
```
[' -e /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-mniS05/anat/sub-mniS05_T2w_RPI_r_labels-disc-manual.nii.gz ']'
+ echo 'manual labeled file was found: /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-mniS05/anat/sub-mniS05_T2w_RPI_r_labels-disc-manual.nii.gz'
manual labeled file was found: /home/paul/Github/csa-atrophy/data-multi-subject/derivatives/labels/sub-mniS05/anat/sub-mniS05_T2w_RPI_r_labels-disc-manual.nii.gz
```

PaulBautin replied:
I could also check if a manual labeling was resampled but not reoriented. This could be a little "heavy" and not really useful in our case.

jcohenadad replied:
no that's too heavy. Why not simply relying only on:
- Lucas's manual label (if exist, you reorient and resample)
- No label (in that case, we do as usual: https://github.com/sct-pipeline/csa-atrophy/blob/f17bd775917e6be4e3ee2e22ea839ecd1455973f/process_data.sh#L67)

PaulBautin replied:
@jcohenadad, I understand now what you meant with :

> if you act on this file now, you probably need to update `label_if_does_not_exist()`

Implemented in https://github.com/sct-pipeline/csa-atrophy/pull/60/commits/d90dc27c6ba6a0c2f2a1da4edb992825ff476563

jcohenadad commented on process_data.sh:
```diff
@@ -54,9 +55,12 @@ label_if_does_not_exist(){
-  FILELABELMANUAL="${PATH_DATA}/derivatives/labels/${SUBJECT}/anat/${FILELABEL}-manual.nii.gz"
-  if [ -e "$FILELABELMANUAL" ]; then
-    echo "manual labeled file was found: $FILELABELMANUAL"
+  FILELABELMANUAL="${path_derivatives}/${file%%_RPI*}_labels-disc-manual.nii.gz"
+  if [ -e "${FILELABELMANUAL}" ]; then
+    echo "manual labeled file was found: ${FILELABELMANUAL##*/}"
```
could you please add a comment explaining the reasons for the %% and *, ie: what file name variants are you expecting to get?  why not just looking for `${file}_labels-disc-manual.nii.gz`?

obviously, $file would not work here (because the variable has been updated), but you can instead input `${SUBJECT}_${contrast_str}`.

PaulBautin replied:
I went with the input `${SUBJECT}_${contrast_str}` in commit https://github.com/sct-pipeline/csa-atrophy/pull/60/commits/b98cacc859e9815821496a66d8e5181c4dc6f5ba

jcohenadad commented on process_data.sh:
```diff
@@ -54,9 +55,12 @@ label_if_does_not_exist(){
-    echo "manual labeled file was found: $FILELABELMANUAL"
+  FILELABELMANUAL="${path_derivatives}/${SUBJECT}_${contrast_str}_labels-disc-manual.nii.gz"
+  if [ -e "${FILELABELMANUAL}" ]; then
+    echo "manual labeled file was found: ${FILELABELMANUAL##*/}"
```
```suggestion
    echo "manual labeled file was found: ${FILELABELMANUAL}"
```

jcohenadad commented on process_data.sh:
```diff
@@ -54,9 +55,12 @@ label_if_does_not_exist(){
+  if [ -e "${FILELABELMANUAL}" ]; then
+    echo "manual labeled file was found: ${FILELABELMANUAL##*/}"
+    # reorienting and resampling image
+    sct_image -i ${FILELABELMANUAL} -setorient RPI -o "${path_derivatives}/${SUBJECT}_${contrast_str}_RPI_labels-disc-manual.nii.gz"
+    sct_resample -i ${FILELABELMANUAL} -mm $interp -o "${path_derivatives}/${SUBJECT}_${contrast_str}_RPI_r_labels-disc-manual.nii.gz"
```
be careful here: you are writing output files into the input (PATH_DATA) folder, which is a bad habit. You'd rather want to copy the derivatives into `$PATH_DATA_PROCESSED` (which is what it was created for)

PaulBautin replied:
Implemented in https://github.com/sct-pipeline/csa-atrophy/pull/60/commits/388172ee37f32b401f94bd7f873514ba83cdc05b

jcohenadad commented:
Same code that could be executed after the if case:
https://github.com/sct-pipeline/csa-atrophy/blob/b98cacc859e9815821496a66d8e5181c4dc6f5ba/process_data.sh#L68
https://github.com/sct-pipeline/csa-atrophy/blob/b98cacc859e9815821496a66d8e5181c4dc6f5ba/process_data.sh#L73

PaulBautin commented:
@jcohenadad, could you re-review when you have a moment?

jcohenadad commented on process_data.sh:
```diff
@@ -94,6 +95,9 @@ segment_if_does_not_exist(){
 # Display useful info for the log, such as SCT version, RAM and CPU cores available
 sct_check_dependencies -short
+# copy derivatives directory containing manual corrections to PATH_DATA_PROCESSED
+cp -r "${PATH_DATA}/derivatives" ${PATH_DATA_PROCESSED}
```
this will be copied every time the script is ran (ie: multiple times, in parallel, with parallel processing, which is unnecessary and prone to simultaneous writing resulting in corruption)-- just copy what you need at the moment, ie: the labels-disc-manual

PaulBautin replied:
Right, https://github.com/sct-pipeline/csa-atrophy/pull/60/commits/070f610a4c2a7334876cac4e4b740be7e23139af corrects this.

jcohenadad commented on process_data.sh:
```diff
@@ -94,6 +95,10 @@ segment_if_does_not_exist(){
 # Display useful info for the log, such as SCT version, RAM and CPU cores available
 sct_check_dependencies -short
+# copy derivatives directory containing manual corrections to PATH_DATA_PROCESSED
+mkdir -p "${PATH_DATA_PROCESSED}/derivatives/labels/${SUBJECT}/anat/"
```
will it crash (exit with non-zero status) if folder exists?

jcohenadad replied:
i just made the test-- it doesn't crash (at least on macOS)

PaulBautin replied:
Also tested on linux,

jcohenadad commented:
@PaulBautin everything looks good to me-- i will launch a process on compute canada now

jcohenadad commented:
i'm just realizing that this branch is not up-to-date with master-- see eg https://github.com/sct-pipeline/csa-atrophy/blob/manual_correction_naming/run_all.py#L28, which is now different in master.

i let you rebase or merge before i run the script

jcohenadad commented:
@PaulBautin what did you do??? i was supposed to test this branch before you merge it! i asked you to update your branch with master (ie merge master --> your branch), not the other way around
